operator/api/v1alpha1/istio_structs.go

```diff
@@ -0,0 +1,170 @@
+	// +kubebuilder:validation:Optional
+	NumTrustedProxies int `json:"numTrustedProxies,omitempty"`
+}
+
+type MeshConfig struct {
+	// Mesh configurion.
+	// +kubebuilder:validation:Optional
+	GatewayTopology GatewayTopology `json:"gatewayTopology,omitempty"`
+}
+
+type ResourceMetricSource struct {
+	// name is the name of the resource in question.
+	// +kubebuilder:validation:Optional
+	Name string `json:"name,omitempty"`
+
+	// targetAverageUtilization is the target value of the average of the
+	// resource metric across all relevant pods, represented as a percentage of
+	// the requested value of the resource for the pods.
+	// +kubebuilder:validation:Optional
+	TargetAverageUtilization int `json:"targetAverageUtilization,omitempty"`
+}
+
+type MetricSpec struct {
+	// type is the type of metric source.  It should be one of "ContainerResource", "External",
+	// "Object", "Pods" or "Resource", each mapping to a matching field in the object.
+	// +kubebuilder:validation:Optional
+	Type string `json:"type,omitempty"`
+
+	// resource refers to a resource metric (such as those specified in
+	// requests and limits) known to Kubernetes describing each pod in the
+	// current scale target (e.g. CPU or memory). Such metrics are built in to
+	// Kubernetes, and have special scaling options on top of those available
+	// to normal per-pod metrics using the "pods" source.
+	// +kubebuilder:validation:Optional
+	Resource ResourceMetricSource `json:"resource,omitempty"`
+}
+
+type HorizontalPodAutoscalerSpec struct {
+	// minReplicas is the lower limit for the number of replicas to which the autoscaler
+	// can scale down. It defaults to 1 pod. minReplicas is allowed to be 0 if the
+	// alpha feature gate HPAScaleToZero is enabled and at least one Object or External
+	// metric is configured. Scaling is active as long as at least one metric value is
+	// available.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Minimum=1
+	MinReplicas int `json:"minReplicas,omitempty"`
+
+	// maxReplicas is the upper limit for the number of replicas to which the autoscaler can scale up.
+	// It cannot be less that minReplicas.
+	// +kubebuilder:validation:Optional
+	MaxReplicas int `json:"maxReplicas,omitempty"`
+
+	// metrics contains the specifications for which to use to calculate the
+	// desired replica count (the maximum replica count across all metrics will
+	// be used). The desired replica count is calculated multiplying the
+	// ratio between the target value and the current value by the current
+	// number of pods. Ergo, metrics used must decrease as the pod count is
+	// increased, and vice-versa. See the individual metric source types for
+	// more information about how each type of metric must respond.
+	// If not set, the default metric will be set to 80% average CPU utilization.
+	// +kubebuilder:validation:Optional
+	Metrics []MetricSpec `json:"metrics,omitempty"`
+}
+
+type RollingUpdateDeployment struct {
```

Comment:
For me to better understand how we want to build our api: Why do we define our own types instead of relying on the k8s api types like `k8s.io.api.apps.v1.RollingUpdateDeployment`? Do we expect to have more restrictions?

Reply:
I would say this gives us better customisability in terms that the entire APIRule has a stable not changing interface. This is also the way Istio configures their CRD in [IstioOperator](https://github.com/istio/operator/blob/master/pkg/apis/istio/v1alpha1/values_types.pb.go), so I would say that is the way to go. @videlov What do you think?

Reply:
Okay agree, full custom. I though we could keep it like k8s api types, so it is more clear for the exact meaning of the type definition, either way is fine with me. Will rework it in the follow-up task for the CRD type definitions.

Reply:
I think we can also discuss this as part of the proposal for the CR with the whole team. It might also be relevant to consider it from a higher perspective, e.g. have some convention between all Kyma modules, even if this introduces more coupling.